cryptol/readme.md

```diff
@@ -0,0 +1,188 @@
+This still won't compile, however, as the Cryptol interpreter will complain about needing certain constraints:
+
+```
+[error]
+  Failed to validate user-specified signature.
+    in the definition of 'leftpad::leftpad'
+    we need to show that
+      for any type pad, in, out
+      the following constraints hold:
+        • fin in
+            arising from
+            use of partial type function (-)
+        • fin pad
+            arising from
+            use of expression (#)
+        • pad >= in
+            arising from
+            use of partial type function (-)
+        • out == pad
+            arising from
+            matching types
+```
+
+Since we use the types `pad`, `in`, and `out` to specify lengths of strings and in functions requiring finite values, we must mark them finite with the `fin` constraint.
+Moreover, because we used `pad - in` as the length of the prefix, it must be non-negative; hence `pad` must be greater than or equal to `in`.
+Finally, the interpreter deduced that the length `out` must be the same as `pad`, which we can check for ourselves from our desired properties of left pad: we need `out` to be the maximum of `pad` and `in`, but `pad >= in` means that `pad` _is_ the maximum of `pad` and `in`.
+
+For ease of reuse, let's factor our the length constraints, and then use them in our definition of `leftpad`.
+
+```cryptol
+type constraint lpCons pad in out = (fin pad, fin in, fin out, pad >= in, out == pad)
+
+leftpad : {pad, in, out} (lpCons pad in out) => Char -> String in -> String out
+leftpad c s = repeat `{pad - in} c # s
+```
+
+With this, asking `cryptol` to load the file completes without complaint:
+
+```
+$ cryptol
+┏━╸┏━┓╻ ╻┏━┓╺┳╸┏━┓╻
+┃  ┣┳┛┗┳┛┣━┛ ┃ ┃ ┃┃
+┗━╸╹┗╸ ╹ ╹   ╹ ┗━┛┗━╸
+version 2.13.0
+https://cryptol.net  :? for help
+
+Loading module Cryptol
+Cryptol> :l leftpad.cry
+Loading module Cryptol
+Loading module leftpad
+```
+
+Now for the properties!
+I'll include the first one here for our discussion; what follows will also apply to the other two properties as well.
+
+Here's the first property in all its glory, reusing our constraint from above:
+
+```cryptol
+leftpadLength : {pad, in, out} (lpCons pad in out) => Char -> String in -> Bit
+property leftpadLength c s = length (leftpad `{pad, in, out} c s) == `out
+```
+
+Recalling the constraints `lpCons`, this says that the length of our output string is `out` A.K.A. `pad` A.K.A. `max(pad, in)`.
+
+Trying to `:prove` this property, however, the interpreter complains:
+
+```
+leftpad> :prove leftpadLength
+Not a monomorphic type:
+{pad, in, out}
+  (fin pad, fin in, fin out, pad >= in, out == pad) =>
+  [8] -> [in][8] -> Bit
+(Total Elapsed Time: 0.000s)
+```
+
+Because the type `{pad, in, out}` is not monomorphic (i.e. instantiated with specific numbers in this case), `cryptol` refuses to try and prove `leftpadLength`.
```

Comment:
To clarify, is this instantiated with specific "number values" or "number types"?

Reply:
Number values at the type level, I think; e.g., we need to say "`pad` refers to the finite number 10" instead of "`pad` is of type `uint64_t`."